src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationKeyComparer.cs

```diff
@@ -29,49 +29,111 @@ public class ConfigurationKeyComparer : IComparer<string>
         /// <returns>Less than 0 if x is less than y, 0 if x is equal to y and greater than 0 if x is greater than y.</returns>
         public int Compare(string? x, string? y)
         {
-            string[] xParts = x?.Split(_keyDelimiterArray, StringSplitOptions.RemoveEmptyEntries) ?? Array.Empty<string>();
-            string[] yParts = y?.Split(_keyDelimiterArray, StringSplitOptions.RemoveEmptyEntries) ?? Array.Empty<string>();
+            ReadOnlySpan<char> xSpan = x.AsSpan();
+            ReadOnlySpan<char> ySpan = y.AsSpan();
+
+            if (x == null || y == null)
+            {
+                return (x is null ? 0 : CountPartsIn(xSpan)) - (y is null ? 0 : CountPartsIn(ySpan));
+            }
 
             // Compare each part until we get two parts that are not equal
-            for (int i = 0; i < Math.Min(xParts.Length, yParts.Length); i++)
+            while (!xSpan.IsEmpty && !ySpan.IsEmpty)
             {
-                x = xParts[i];
-                y = yParts[i];
+                if (xSpan[0] == s_keyDelimiter)
+                {
+                    xSpan = xSpan.Slice(1);
+                    continue;
+                }
 
-                int value1 = 0;
-                int value2 = 0;
+                if (ySpan[0] == s_keyDelimiter)
+                {
+                    ySpan = ySpan.Slice(1);
+                    continue;
+                }
 
-                bool xIsInt = x != null && int.TryParse(x, out value1);
-                bool yIsInt = y != null && int.TryParse(y, out value2);
+                int xSegmentLength = xSpan.IndexOf(s_keyDelimiter);
+                if (xSegmentLength < 0)
+                {
+                    xSegmentLength = xSpan.Length;
+                }
+
+                int ySegmentLength = ySpan.IndexOf(s_keyDelimiter);
+                if (ySegmentLength < 0)
+                {
+                    ySegmentLength = ySpan.Length;
+                }
+
+                int compareResult = Compare(xSpan.Slice(0, xSegmentLength), ySpan.Slice(0, ySegmentLength));
+                if (compareResult != 0)
+                {
+                    return compareResult;
+                }
+
+                xSpan = xSpan.Slice(xSegmentLength);
+                ySpan = ySpan.Slice(ySegmentLength);
+            }
 
+            if (xSpan.IsEmpty)
+            {
+                return ySpan.IsEmpty ? 0 : -CountPartsIn(ySpan);
```

Comment:
With the above suggestion:
```suggestion
                return -CountPartsIn(ySpan);
```
So when `ySpan.IsEmpty` `0` is returned, hence the `-` doesn't bother us.